### ISSUE TITLE
Javascript implementation for Keccak/SHAKE/K12

### ISSUE THREAD

**Issue by DonaldTsang** [XKCP/XKCP#47]
Missing base implementations for browsers.
- Needs int64 and shitfting
- Needs arrays of bytes with XORing
- Needs right-encoding

**Comment by gvanas** [comment]
Hi,

Thanks for your comment but could you make your request/question more concrete?

For a Javascript implementation, did you have a look at [js-sha3](https://github.com/emn178/js-sha3)?

Kind regards,
Gilles

**Comment by DonaldTsang** [comment]
@gvanas Last time I checked JS-SHA3 is missing KangarooTwelve and MarsupilamiFourteen, so it would be great if JS-SHA3 or other JS libraries to have those two XOF as well.

**Comment by gvanas** [comment]
I think it would be a good idea indeed to ask the authors of JS-SHA3 to add K12, or to contribute to it (but I won't volunteer as I don't know Javascript).

**Comment by DonaldTsang** [comment]
Asked https://github.com/emn178/js-sha3/issues/18

**Comment by DonaldTsang** [comment]
@gvanas by comparing these two pieces of source code:
https://github.com/XKCP/XKCP/blob/master/Standalone/CompactFIPS202/Python/CompactFIPS202.py
https://github.com/XKCP/XKCP/blob/master/Standalone/KangarooTwelve/Python/K12.py
The difference between these two is just differences between `KeccakP1600` vs `KeccakF1600` ad everything that supersedes those two functions, so `KeccakP1600onLanes`  and other one-line functions are untouched

**Comment by gvanas** [comment]
Yes, once you have an implementation of Keccak (or SHA-3 or SHAKE), it is fairly easy to adapt it to KangarooTwelve indeed.

**Comment by DonaldTsang** [comment]
@gvanas I see some notable differences that I can't quite understand.
For CompactFIPS202,
```    state = bytearray(200)
    for x in range(5):
        for y in range(5):
            state[8*(x+5*y):8*(x+5*y)+8] = store64(lanes[x][y])
    return state
```
And for KangarooTwelve,
```
    state = bytearray().join([store64(lanes[x][y]) for y in range(5) for x in range(5)])
    return bytearray(state)
```
Are they functional equivalents?

**Comment by DonaldTsang** [comment]
@gvanas the latter is cleaner TBH, one-liners are sweet if they are structual

**Comment by d3x0r** [comment]
Did a wasm compilation of https://github.com/XKCP/K12 with a small JS interface wrapper... it's not great.

https://github.com/d3x0r/K12#wasm-compilation

**Comment by DonaldTsang** [comment]
Thx!
